test/common/verify_alert_links_to_SOPs.go

```diff
@@ -0,0 +1,147 @@
+package common
+
+import (
+	"encoding/json"
+	"fmt"
+	prometheusv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+)
+
+var githubToken = os.Getenv("TOKEN_GITHUB")
+
+func TestVerifyAlertLinksInSOPs(t TestingTB, ctx *TestingContext) {
+
+	if githubToken == "" {
+		t.Skip("github token not present")
+	}
+
+	var apiLinks []string
+	testUrl := "https://github.com/RHCloudServices/integreatly-help/blob/master/sops/README.md"
+	testResp := getSOPAlertLinkStatus(t, testUrl)
+
+	if testResp.StatusCode != 200 {
+		t.Log("Response status: ", testUrl, testResp.Status)
+	} else {
+		t.Log("Response status: ", testResp.Status)
+	}
+
+	output, err := execToPod("wget -qO - localhost:9090/api/v1/rules",
+		"prometheus-prometheus-0",
+		ObservabilityProductNamespace,
+		"prometheus", ctx)
+	if err != nil {
+		t.Fatal("failed to exec to pod:", err)
+	}
+
+	var ApiOutput prometheusAPIResponse
+
+	err = json.Unmarshal([]byte(output), &ApiOutput)
+	if err != nil {
+		t.Fatal("Failed to unmarshal json", err)
+	}
+
+	var rulesResult prometheusv1.RulesResult
+
+	err = json.Unmarshal([]byte(ApiOutput.Data), &rulesResult)
+	if err != nil {
+		t.Fatal("Failed to unmarshal json", err)
+	}
+
+	for _, group := range rulesResult.Groups {
+		for _, rule := range group.Rules {
+			switch v := rule.(type) {
+			case prometheusv1.RecordingRule:
+			case prometheusv1.AlertingRule:
+				for annotation, sopUrl := range v.Annotations {
+					if annotation == "sop_url" {
+						apiLinks = append(apiLinks, string(sopUrl))
+					}
+
+				}
+
+			default:
+				fmt.Printf("Unknown rule type %s", v)
+
+			}
+		}
+	}
+
+	apiLinks = unique(apiLinks)
+	result := validateSOPLinks(t, apiLinks)
+	fmt.Println(result)
+}
+
+func modifyLink(rawLink string) (url string) {
+	r := strings.NewReplacer(
+		"github", "api.github",
+		"com/", "com/repos/",
+		"blob/master", "contents",
+		"tree/master", "contents",
+	)
+
+	url = r.Replace(rawLink)
+	return url
+}
+
+func unique(s []string) []string {
+	inResult := make(map[string]bool)
+	var result []string
+	for _, str := range s {
+		if _, ok := inResult[str]; !ok {
+			inResult[str] = true
+			result = append(result, str)
+		}
+	}
+	return result
+}
+
+func validateSOPLinks(t TestingTB, s []string) bool {
+	status := true
+	var countFailedLinks int
+	for _, url := range s {
+		resp := getSOPAlertLinkStatus(t, url)
+
+		if resp.StatusCode != 200 {
+			t.Log("Response status :", resp.Status)
+			t.Log(url)
+			countFailedLinks++
+		}
+
+		if countFailedLinks != 0 {
+			status = false
+		}
```

Comment:
The whole `if` should be moved outside the `for` loop. Reason is that in `for` loop you just count the failed links but it is enough to check the `countFailedLinks` just once once the `for` loop is over.

Reply:
Good point. Thank you